File: android2/app/src/main/java/com/xbot/xbot/face/ExpressionClassifier.java
package com.xbot.xbot.face;

import com.xbot.xbot.model.Expression;
import com.xbot.xbot.model.ExpressionResult;

import java.util.EnumMap;
import java.util.Map;

/**
 * Rule-based expression classifier from ARKit 52 blendshape coefficients.
 *
 * <p>Ported from Flutter {@code lib/services/expression_classifier.dart}.
 */
public class ExpressionClassifier {
    private static final String MOUTH_SMILE_LEFT = "mouthSmileLeft";
    private static final String MOUTH_SMILE_RIGHT = "mouthSmileRight";
    private static final String CHEEK_SQUINT_LEFT = "cheekSquintLeft";
    private static final String CHEEK_SQUINT_RIGHT = "cheekSquintRight";
    private static final String MOUTH_FROWN_LEFT = "mouthFrownLeft";
    private static final String MOUTH_FROWN_RIGHT = "mouthFrownRight";
    private static final String BROW_INNER_UP = "browInnerUp";
    private static final String BROW_OUTER_UP_LEFT = "browOuterUpLeft";
    private static final String BROW_OUTER_UP_RIGHT = "browOuterUpRight";
    private static final String BROW_DOWN_LEFT = "browDownLeft";
    private static final String BROW_DOWN_RIGHT = "browDownRight";
    private static final String JAW_OPEN = "jawOpen";
    private static final String EYE_WIDE_LEFT = "eyeWideLeft";
    private static final String EYE_WIDE_RIGHT = "eyeWideRight";
    private static final String EYE_SQUINT_LEFT = "eyeSquintLeft";
    private static final String EYE_SQUINT_RIGHT = "eyeSquintRight";
    private static final String NOSE_SNEER_LEFT = "noseSneerLeft";
    private static final String NOSE_SNEER_RIGHT = "noseSneerRight";
    private static final String MOUTH_UPPER_UP_LEFT = "mouthUpperUpLeft";
    private static final String MOUTH_UPPER_UP_RIGHT = "mouthUpperUpRight";
    private static final String MOUTH_STRETCH_LEFT = "mouthStretchLeft";
    private static final String MOUTH_STRETCH_RIGHT = "mouthStretchRight";
    private static final String MOUTH_PRESS_LEFT = "mouthPressLeft";
    private static final String MOUTH_PRESS_RIGHT = "mouthPressRight";

    private final double activationThreshold;

    public ExpressionClassifier() {
        this(0.28);
    }

    public ExpressionClassifier(double activationThreshold) {
        this.activationThreshold = activationThreshold;
    }

    public ExpressionResult classify(Map<String, Double> blendshapes) {
        if (blendshapes == null || blendshapes.isEmpty()) {
            return ExpressionResult.NEUTRAL;
        }

        double smile = avg(blendshapes, MOUTH_SMILE_LEFT, MOUTH_SMILE_RIGHT);
        double cheekSquint = avg(blendshapes, CHEEK_SQUINT_LEFT, CHEEK_SQUINT_RIGHT);
        double frown = avg(blendshapes, MOUTH_FROWN_LEFT, MOUTH_FROWN_RIGHT);
        double browInnerUp = v(blendshapes, BROW_INNER_UP);
        double browOuterUp = avg(blendshapes, BROW_OUTER_UP_LEFT, BROW_OUTER_UP_RIGHT);
        double browDown = avg(blendshapes, BROW_DOWN_LEFT, BROW_DOWN_RIGHT);
        double jawOpen = v(blendshapes, JAW_OPEN);
        double eyeWide = avg(blendshapes, EYE_WIDE_LEFT, EYE_WIDE_RIGHT);
        double eyeSquint = avg(blendshapes, EYE_SQUINT_LEFT, EYE_SQUINT_RIGHT);
        double noseSneer = avg(blendshapes, NOSE_SNEER_LEFT, NOSE_SNEER_RIGHT);
        double upperLipUp = avg(blendshapes, MOUTH_UPPER_UP_LEFT, MOUTH_UPPER_UP_RIGHT);
        double mouthStretch = avg(blendshapes, MOUTH_STRETCH_LEFT, MOUTH_STRETCH_RIGHT);
        double mouthPress = avg(blendshapes, MOUTH_PRESS_LEFT, MOUTH_PRESS_RIGHT);

        Map<Expression, Double> scores = new EnumMap<>(Expression.class);
        scores.put(Expression.HAPPY, smile * 0.8 + cheekSquint * 0.2);
        scores.put(Expression.SAD, frown * 0.6 + browInnerUp * 0.4 - smile * 0.3);
        scores.put(Expression.SURPRISED,
                jawOpen * 0.45 + browOuterUp * 0.25 + browInnerUp * 0.15 + eyeWide * 0.15);
        scores.put(Expression.ANGRY,
                browDown * 0.6 + mouthPress * 0.2 + eyeSquint * 0.2 - jawOpen * 0.2);
        scores.put(Expression.DISGUSTED, noseSneer * 0.6 + upperLipUp * 0.4);
        scores.put(Expression.FEARFUL, eyeWide * 0.4 + browInnerUp * 0.3 + mouthStretch * 0.3);

        Expression best = Expression.NEUTRAL;
        double bestScore = 0;
        for (Map.Entry<Expression, Double> entry : scores.entrySet()) {
            double clamped = clamp(entry.getValue());
            if (clamped > bestScore) {
                bestScore = clamped;
                best = entry.getKey();
            }
        }

        Map<Expression, Double> fullScores = new EnumMap<>(Expression.class);
        fullScores.put(Expression.NEUTRAL, Math.max(0.0, 1.0 - bestScore));
        for (Map.Entry<Expression, Double> entry : scores.entrySet()) {
            fullScores.put(entry.getKey(), clamp(entry.getValue()));
        }

        if (bestScore < activationThreshold) {
            Double neutralScore = fullScores.get(Expression.NEUTRAL);
            return new ExpressionResult(
                    Expression.NEUTRAL,
                    neutralScore != null ? neutralScore : 1.0,
                    fullScores);
        }

        return new ExpressionResult(best, bestScore, fullScores);
    }

    private static double v(Map<String, Double> b, String key) {
        Double value = b.get(key);
        return clamp(value != null ? value : 0);
    }

    private static double avg(Map<String, Double> b, String a, String c) {
        return (v(b, a) + v(b, c)) / 2.0;
    }

    private static double clamp(double value) {
        return Math.max(0.0, Math.min(1.0, value));
    }
}

File: xbot-native/app/src/main/java/com/xbot/android/detection/ExpressionClassifier.kt
package com.xbot.android.detection

import com.xbot.android.model.Expression
import com.xbot.android.model.ExpressionResult

/**
 * 规则引擎表情分类:从 ARKit 52 blendshapes 计算 7 种表情分数。
 * 直接翻译自 Flutter 的 ExpressionClassifier。无训练模型,纯数学。
 */
class ExpressionClassifier(
    private val activationThreshold: Float = 0.28f
) {

    /**
     * 从 blendshape 系数计算表情。
     * @param blendshapes MediaPipe FaceLandmarker 输出的 blendshape Map(key=名称, value=0..1)
     */
    fun classify(blendshapes: Map<String, Float>): ExpressionResult {
        val happy = (blendshapes["mouthSmileLeft"] ?: 0f) * 0.8f +
                (blendshapes["cheekSquintLeft"] ?: 0f) * 0.2f

        val sad = (blendshapes["mouthFrownLeft"] ?: 0f) * 0.6f +
                (blendshapes["browInnerUp"] ?: 0f) * 0.4f -
                (blendshapes["mouthSmileLeft"] ?: 0f) * 0.3f

        val surprised = (blendshapes["jawOpen"] ?: 0f) * 0.45f +
                (blendshapes["browOuterUpLeft"] ?: 0f) * 0.25f +
                (blendshapes["browInnerUp"] ?: 0f) * 0.15f +
                (blendshapes["eyeWideLeft"] ?: 0f) * 0.15f

        val angry = (blendshapes["browDownLeft"] ?: 0f) * 0.6f +
                (blendshapes["mouthPressLeft"] ?: 0f) * 0.2f +
                (blendshapes["eyeSquintLeft"] ?: 0f) * 0.2f -
                (blendshapes["jawOpen"] ?: 0f) * 0.2f

        val disgusted = (blendshapes["noseSneerLeft"] ?: 0f) * 0.6f +
                (blendshapes["upperLipUpLeft"] ?: 0f) * 0.4f

        val fearful = (blendshapes["eyeWideLeft"] ?: 0f) * 0.4f +
                (blendshapes["browInnerUp"] ?: 0f) * 0.3f +
                (blendshapes["mouthStretchLeft"] ?: 0f) * 0.3f

        val scores = mapOf(
            Expression.HAPPY to happy.coerceIn(0f, 1f),
            Expression.SAD to sad.coerceIn(0f, 1f),
            Expression.SURPRISED to surprised.coerceIn(0f, 1f),
            Expression.ANGRY to angry.coerceIn(0f, 1f),
            Expression.DISGUSTED to disgusted.coerceIn(0f, 1f),
            Expression.FEARFUL to fearful.coerceIn(0f, 1f)
        )

        val winner = scores.maxByOrNull { it.value }!!
        return if (winner.value >= activationThreshold) {
            ExpressionResult(winner.key, winner.value, scores)
        } else {
            ExpressionResult(Expression.NEUTRAL, 0f, scores)
        }
    }
}
